File: src/main/java/com/hp/jipp/model/JobState.java
package com.hp.jipp.model;

import com.google.auto.value.AutoValue;
import com.hp.jipp.encoding.NameCodeEncoder;
import com.hp.jipp.encoding.NameCode;

/**
 * Job State values.
 *
 * @see <a href="https://tools.ietf.org/html/rfc2911#section-4.3.7">RFC2911 Section 4.3.7</a>
 */
@AutoValue
public abstract class JobState extends NameCode {

    public static final JobState Pending = of("pending", 3);
    public static final JobState PendingHeld = of("pending-held", 4);
    public static final JobState Processing = of("processing", 5);
    public static final JobState ProcessingStopped = of("processing-stopped", 6);
    public static final JobState Canceled = of("canceled", 7);
    public static final JobState Aborted = of("aborted", 8);
    public static final JobState Completed = of("completed", 9);

    public static final NameCodeEncoder<JobState> ENCODER = NameCodeEncoder.of(
            JobState.class, new NameCode.Factory<JobState>() {
                @Override
                public JobState of(String name, int code) {
                    return JobState.of(name, code);
                }
            });

    /** Return true if this state is the terminating state for a job */
    public boolean isFinal() {
        return this == Canceled || this == Aborted || this == Completed;
    }

    public static JobState of(String name, int code) {
        return new AutoValue_JobState(name, code);
    }
}

File: src/main/java/com/hp/jipp/model/JobState.kt
// Copyright 2017 HP Development Company, L.P.
// SPDX-License-Identifier: MIT

package com.hp.jipp.model

import com.hp.jipp.encoding.Enum
import com.hp.jipp.encoding.EnumType
import com.hp.jipp.encoding.encoderOf

/**
 * Job State values.

 * See [RFC2911 Section 4.3.7](https://tools.ietf.org/html/rfc2911.section-4.3.7)
 */
data class JobState( override val code: Int, override val name: String) : Enum() {

    /** Return true if this state is the terminating state for a job  */
    val isFinal: Boolean
        get() = this == canceled || this == aborted || this == completed

    /** The attribute type for a [JobState] attribute */
    class Type(name: String) : EnumType<JobState>(ENCODER, name)

    companion object {
        @JvmField val pending = JobState(3, "pending")
        @JvmField val pendingHeld = JobState(4, "pending-held")
        @JvmField val processing = JobState(5, "processing")
        @JvmField val processingStopped = JobState(6, "processing-stopped")
        @JvmField val canceled = JobState(7, "canceled")
        @JvmField val aborted = JobState(8, "aborted")
        @JvmField val completed = JobState(9, "completed")

        @JvmField val ENCODER = encoderOf(JobState::class.java, { code, name -> JobState(code, name) })
    }
}
